Add KDE Plasma desktop environment support

joshua1955 opened the pull request:
Add support for changing wallpapers in KDE Plasma using the plasma-apply-wallpaperimage command.

Changes:
- Add set_plasma_wallpaper() function in chwall/wallpaper.py
- Add 'KDE Plasma' option to desktop environment selector in GUI
- Update configuration example with plasma desktop option
- Update README to list KDE Plasma as supported environment
- Add detailed configuration documentation in CONFIG.org

Users can now select KDE Plasma from the preferences GUI or set 'desktop: plasma' in their chwall.yml configuration file to use plasma-apply-wallpaperimage for wallpaper changes.

joshua1955 committed:
Add KDE Plasma desktop environment support

Add support for changing wallpapers in KDE Plasma using the
plasma-apply-wallpaperimage command.

Changes:
- Add set_plasma_wallpaper() function in chwall/wallpaper.py
- Add 'KDE Plasma' option to desktop environment selector in GUI
- Update configuration example with plasma desktop option
- Update README to list KDE Plasma as supported environment
- Add detailed configuration documentation in CONFIG.org

Users can now select KDE Plasma from the preferences GUI or set
'desktop: plasma' in their chwall.yml configuration file to use
plasma-apply-wallpaperimage for wallpaper changes.

joshua1955 committed:
Add KDE Plasma lock screen wallpaper synchronization support

This commit introduces the ability to automatically sync KDE Plasma lock screen
wallpapers with desktop wallpapers when they change.

1. **New Functionality (wallpaper.py)**
   - Added `set_kde_lock_wallpaper()` function to manage KDE lock screen settings
   - Updates `Image` and `PreviewImage` parameters in `~/.config/kscreenlockerrc`
   - Integrated with existing `set_wallpaper()` for automatic synchronization

2. **Configuration Support (utils.py)**
   - Added new `lock_screen_sync` setting in general configuration
   - Defaults to False to maintain backward compatibility

3. **GUI Integration (preferences.py)**
   - Added toggle switch in General preferences section
   - Label: "Sync lock screen wallpaper with desktop wallpaper (KDE Plasma)"
   - Automatically saves to configuration when changed

- Available only when KDE Plasma desktop environment is selected
- Users can enable/disable synchronization via preferences
- Lock screen wallpaper updates automatically when desktop wallpaper changes
- No impact on existing functionality for other desktop environments

milouse commented on chwall/gui/preferences.py:
```diff
@@ -568,6 +568,7 @@ def make_general_pane(self):
         environments = [("gnome", "Gnome, Pantheon, Budgie, …"),
                         ("mate", "Mate"),
                         ("xfce", "XFCE"),
+                        ("plasma", "KDE Plasma"),
```
You can move this value in second position, between gnome and mate :)

joshua1955 replied:
Ok, thx :)

milouse commented on chwall/wallpaper.py:
```diff
@@ -224,6 +224,66 @@ def set_gnome_screensaver(path):
+        image_found = False
+        preview_found = False
+        
+        for i, line in enumerate(config_content):
```
From what I read, kscreenlockerrc looks like an INI file. Can you rewrite this part by using directly [configparser](https://docs.python.org/3/library/configparser.html) ?

joshua1955 replied:
Check my new commit, plz :)

milouse commented on chwall/wallpaper.py:
```diff
@@ -252,6 +312,15 @@ def set_wallpaper(path, config):
         globals()[screensaver_method](shared_path or path)
+    
+    # Set KDE lock screen wallpaper if enabled
+    if (desktop == "plasma" and 
```
Even if it seems very unlikely, I think this first check is useless: we might have someone using kscreenlocker without plasma (or the reverse). Thus for me just testing that lock_screen_sync is True is sufficient.

joshua1955 replied:
Removed plasma desktop check.

milouse commented on chwall/gui/preferences.py:
```diff
@@ -592,6 +593,12 @@ def make_general_pane(self):
         genbox.pack_start(prefbox, False, False, 0)
 
+        prefbox = self.make_toggle_pref(
+            "general", "lock_screen_sync",
```
Please, use a more precise variable name, like `kde_screen_locker_sync` (or `plasma…`)

joshua1955 replied:
Changed it in last commit.

milouse commented on chwall.yml.example:
```diff
@@ -8,6 +8,9 @@ bing:
 general:
+  # Desktop environment: gnome, xfce, mate, sway, feh, plasma
+  # For KDE Plasma, use: desktop: plasma
+  # desktop: plasma
```
Line 12 is not necessary. Or if you think people might be confused by the word "plasma", use kde as value everywhere (I have no opinion in the matter as I never ever used KDE myself, so I’m pretty far away from all this wording ^^).

On line 13, if you provide an exemple, it must be the default value. So `desktop: gnome`.

joshua1955 replied:
Check my new commit, plz :)

milouse commented on CONFIG.org:
```diff
@@ -37,6 +41,7 @@ objects:
 #+begin_src yaml
 ---
 general:
+  desktop: plasma  # For KDE Plasma users
```
Non necessary (or we should add an exemple for all other DE… people will find the right value to use).

joshua1955 replied:
Check my new commit, plz :)

milouse commented on README.org:
```diff
@@ -29,7 +29,7 @@ sources:
 
 It currently supports some major desktop environments (Gnome, XFCE,
-Mate) and light window managers (Sway and anything accepting feh).
+Mate, KDE Plasma) and light window managers (Sway and anything accepting feh).
```
Please, ensure that the line is no more than 72 char long (sorry about this implied rules, I should fix that in the editorconfig file).

milouse reviewed (changes requested):
Thank you very much for your contribution. I’m glad to see some real users :)

I left some comments before accepting it, but overall it’s all good, thank you again for your work!

joshua1955 committed:
Fix: Moved 'plasma' to 2-nd position in GUI
Fix: Edit kscreenlockerrc using configparser for stability
Fix: Removed check for 'plasma' desktop in wallpaper sync function
Fix: Changed var 'lock_screen_sync' to 'kde_screen_locker_sync'
Fix: Removed optional desctiption for 'plasma' desktop in CONFIG.org
Fix: Edited description in README.org due to 72 char long

joshua1955 commented:
Hello! I really like your app and found it unfortunately does not work in my KDE environment.

I'm relatively new to working with GitHub, but I'm keen to get some practice and would love to try and contribute to solving this. I'm ready to follow any guidelines for debugging or testing you can provide!

milouse commented:
Hello,

Don’t worry, current workflow is fine :)

On a second think, though, I’ll reject the whole kscreenlockerrc sync code, as this is probably a duplicate of the already existing settings. To achieve the same result, you should set the "Shared backgroud path" to anything (or the key `general.shared.path` in the chwall.yml config file), with or without (as you wish) the "Blur shared background" setting as well.

This shared background path **must be** an existing file. The idea is that each time chwall change the desktop background, it will overwrite this given file with the current background picture (blurred or not). So to begin with, you can just create a blank file anywhere you like with `touch ~/.cache/chwall_locker_bg` for exemple. Like this you will be able to pick it in both the GUI config of chwall and the KDE settings.

Then in KDE settings directly, go to the screenlocker settings and choose the same file path for the background of the locker. Then, as this file will be updated in place each time the background change, no need to update the locker config file.

<img width="1866" height="558" alt="Capture du 2025-11-06 14-26-29" src="https://github.com/user-attachments/assets/9ddaa0d5-0ddc-4154-8964-c13801f36761" />

A possible improvement here being to add the ability to create this shared background file. I understand the current "shared background path" pref is a bit cryptic and hard to set from scratch.

milouse commented:
Merged in https://git.umaneti.net/chwall/commit/?h=v0.7.3-dev&id=17d37a972594633360d98c0b33fbe40d7ec6a5ca

~~I’m trying to push an update to the chwall-git package on aur, but somehow it doesn’t work (probably related to all their issues lately with the ddos attacks). I hope to be able to do it asap.~~ [EDIT]Done… it was just me forgetting to update the `.SRCINFO` file 🤦 You can now try your changes by using the chwall-git package if you wish[/EDIT]

Thanks again for your contribution!
